common/membership/sharddistributorresolver.go

```diff
@@ -0,0 +1,167 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package membership
+
+import (
+	"context"
+
+	"github.com/uber/cadence/client/sharddistributor"
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/types"
+)
+
+type modeKey string
+
+var (
+	modeKeyHashRing                       modeKey = "hash_ring"
+	modeKeyShardDistributor               modeKey = "shard_distributor"
+	modeKeyHashRingShadowShardDistributor modeKey = "hash_ring-shadow-shard_distributor"
+	modeKeyShardDistributorShadowHashRing modeKey = "shard_distributor-shadow-hash_ring"
+)
+
+type shardDistributorResolver struct {
+	namespace             string
+	shardDistributionMode dynamicconfig.StringPropertyFn
+	client                sharddistributor.Client
+	ring                  SingleProvider
+	logger                log.Logger
+}
+
+func (s shardDistributorResolver) AddressToHost(owner string) (HostInfo, error) {
+	return s.ring.AddressToHost(owner)
+}
+
+func NewShardDistributorResolver(
```

Comment:
The name is confusing. At the first glance, I thought it implements `Resolver` interface, but it actually implements `SingleProvider` interface.